dask_cuda/proxy_object.py

```diff
@@ -0,0 +1,483 @@
+
+    Parameters
+    ----------
+    obj: object
+        Object to wrap in a ProxyObject object.
+    serializers: List[Str], optional
+        List of serializers to use to serialize `obj`. If None,
+        no serialization is done.
+    subclass: Class, optional
+        Specify a subclass of ProxyObject to create instead of ProxyObject.
+        `subclass` must be pickable.
+    ret: ProxyObject
+        The proxy object proxing `obj`
+    """
+
+    if hasattr(obj, "_obj_pxy"):  # Already a proxy object
+        ret = obj
+    else:
+        fixed_attr = {}
+        for attr in _FIXED_ATTRS:
+            try:
+                fixed_attr[attr] = getattr(obj, attr)
+            except AttributeError:
+                pass
+
+        if subclass is None:
+            subclass = ProxyObject
+        ret = subclass(
+            obj=obj,
+            fixed_attr=fixed_attr,
+            type_serialized=pickle.dumps(type(obj)),
+            typename=dask.utils.typename(type(obj)),
+            is_cuda_object=is_device_object(obj),
+            subclass=pickle.dumps(subclass) if subclass else None,
+            serializers=None,
+        )
+    if serializers is not None:
+        ret._obj_pxy_serialize(serializers=serializers)
+    return ret
+
+
+class ProxyObject:
+    __slots__ = [
+        "_obj_pxy",  # A dict that holds the state of the proxy object
+        "_obj_pxy_lock",  # Threading lock for all obj_pxy access
+        "__obj_pxy_cache",  # A dict used for caching attributes
+    ]
+
+    def __init__(
+        self,
+        obj,
+        fixed_attr,
+        type_serialized,
+        typename,
+        is_cuda_object,
+        subclass,
+        serializers,
+    ):
+        self._obj_pxy = {
+            "obj": obj,
+            "fixed_attr": fixed_attr,
+            "type_serialized": type_serialized,
+            "typename": typename,
+            "is_cuda_object": is_cuda_object,
+            "subclass": subclass,
+            "serializers": serializers,
+        }
+        self._obj_pxy_lock = threading.RLock()
+        self.__obj_pxy_cache = {}
+
+    def _obj_pxy_get_meta(self):
+        """Return the metadata of the proxy object.
+
+        Returns
+        -------
+        ret: dict
+            Dictionary of metadata
+        """
+        with self._obj_pxy_lock:
+            return {k: self._obj_pxy[k] for k in self._obj_pxy.keys() if k != "obj"}
+
+    def _obj_pxy_serialize(self, serializers):
+        """Inplace serialization of the proxied object using the `serializers`
+
+        Parameters
+        ----------
+        serializers: List[Str]
+            List of serializers to use to serialize the proxied object.
+
+        Returns
+        -------
+        header: dict
+            The header of the serialized frames
+        frames: List[Bytes]
+            List of frames that makes up the serialized object
+        """
+        with self._obj_pxy_lock:
+            assert serializers is not None
```

Comment:
I am not sure, the check is there to assert the internal logic of the class. The `_obj_pxy_*` methods are not supposed to be called by the user.

Reply:
I understand that, but nevertheless I think of `assert` as a debug statement, so I'm not sure it makes much sense in runtime code. I know Python doesn't distinguish between "debug" and "release" builds, like we can with C, but regardless I think it's clearer to have errors be specific, one could also argue that assertions would substitute virtually any exceptions we could raise, but the exceptions let us be clearer about our intent.

In the interest of avoiding nitpicking, I will leave the final decision up to you, but I think there's real value in being more specific instead of using assertions, whenever we want to raise an exception if something goes wrong.

Reply:
I am not sure I agree completely. I like to use `assert` in internal logic in order to both catch bugs while developing and document the expected state of an object. 
However, in this case I agree with you. Since it can be triggered by user input, it is preferable to raise `ValueError`.